cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -354,13 +377,397 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
             return Future.succeededFuture(this);
         }
 
+        /**
+         * If the SS exists, complete any pending rolls
+         *
+         * @return A Future which completes with the current state of the SS, or with null if the SS never existed.
+         */
+        public Future<StatefulSet> waitForQuiescence(String namespace, String statefulSetName) {
+            return kafkaSetOperations.getAsync(namespace, statefulSetName).compose(ss -> {
+                if (ss != null) {
+                    return kafkaSetOperations.maybeRollingUpdate(ss,
+                        pod -> {
+                            boolean notUpToDate = !isPodUpToDate(ss, pod);
+                            if (notUpToDate) {
+                                log.debug("Rolling pod {} prior to upgrade", pod.getMetadata().getName());
+                            }
+                            return notUpToDate;
+                        }).map(ignored -> ss);
+                } else {
+                    return Future.succeededFuture(ss);
+                }
+            });
+        }
+
+        Future<ReconciliationState> kafkaUpgrade() {
+            // Wait until the SS is not being updated (it shouldn't be, but there's no harm in checking)
+            String kafkaSsName = KafkaCluster.kafkaClusterName(name);
+            return waitForQuiescence(namespace, kafkaSsName).compose(
+                ss -> {
+                    if (ss == null) {
+                        return Future.succeededFuture(this);
+                    }
+                    log.debug("Does SS {} need to be upgraded?", ss.getMetadata().getName());
+                    Future<?> result;
+                    // Get the current version of the cluster
+                    KafkaVersion currentVersion = versions.version(ss.getMetadata().getAnnotations().get(KafkaCluster.ANNO_STRIMZI_IO_KAFKA_VERSION));
+                    log.debug("SS {} has current version {}", ss.getMetadata().getName(), currentVersion);
+                    String fromVersionAnno = ss.getMetadata().getAnnotations().get(ANNO_STRIMZI_IO_FROM_VERSION);
+                    KafkaVersion fromVersion;
+                    if (fromVersionAnno != null) { // We're mid-upgrade
+                        fromVersion = versions.version(fromVersionAnno);
+                    } else {
+                        fromVersion = currentVersion;
+                    }
+                    log.debug("SS {} is upgrading from version {}", ss.getMetadata().getName(), fromVersion);
```

Comment:
is "upgrading" the right word? it could be a "downgrade" right? maybe it could be confusing for people who read the log.